tests/unit/run_modes/test_skip_units.py

```diff
@@ -0,0 +1,135 @@
+    """
+    tdef = SimpleNamespace(rtconfig=conf)
+    tdef.name = 'foo'
+    assert check_task_skip_config(tdef) is None
+
+
+def test_raises_check_task_skip_config():
+    """It raises an error if succeeded and failed are set.
+    """
+    tdef = SimpleNamespace(
+        rtconfig={'skip': {'outputs': ['foo1', 'failed', 'succeeded']}}
+    )
+    tdef.name = 'foo'
+    with raises(WorkflowConfigError, match='succeeded AND failed'):
+        check_task_skip_config(tdef)
+
+
+@pytest.mark.parametrize(
+    'outputs, required, expect',
+    (
+        param([], [], ['succeeded'], id='implicit-succeded'),
+        param(
+            ['succeeded'], ['succeeded'], ['succeeded'],
+            id='explicit-succeded'
+        ),
+        param(['submitted'], [], ['succeeded'], id='only-1-submit'),
+        param(
+            ['foo', 'bar', 'baz', 'qux'],
+            ['bar', 'qux'],
+            ['bar', 'qux', 'succeeded'],
+            id='required-only'
+        ),
+        param(
+            ['foo', 'baz'],
+            ['bar', 'qux'],
+            ['succeeded'],
+            id='no-required'
+        ),
+        param(
+            ['failed'],
+            [],
+            ['failed'],
+            id='explicit-failed'
+        ),
+    )
+)
+def test_process_outputs(outputs, required, expect):
+    """Check that skip outputs:
+
+    1. Doesn't send submitted twice.
+    2. Sends every required output.
+    3. If failed is set send failed
+    4. If failed in not set send succeeded.
+    """
+    # Create a mocked up task-proxy:
+    rtconf = {'skip': {'outputs': outputs}}
+    itask = SimpleNamespace(
+        tdef=SimpleNamespace(
+            rtconfig=rtconf),
+        state=SimpleNamespace(
+            outputs=SimpleNamespace(
+                iter_required_messages=lambda exclude: iter(required),
+                _message_to_trigger={v: v for v in required}
+            )))
+
+    assert process_outputs(itask, rtconf) == ['submitted', 'started'] + expect
+
+
+def test_skip_mode_validate(monkeypatch, caplog):
+    """It warns us if we've set a task config to nonlive mode.
+
+    (And not otherwise)
+
+    Point 3 from the skip mode proposal
+    https://github.com/cylc/cylc-admin/blob/master/docs/proposal-skip-mode.md
+
+    | If the run mode is set to simulation or skip in the workflow
+    | configuration, then cylc validate and cylc lint should produce 
+    | warning (similar to development features in other languages / systems).
+    """
+    taskdefs = {
+        f'{run_mode}_task': SimpleNamespace(
+            rtconfig={'run mode': run_mode},
+            name=f'{run_mode}_task'
+        )
+        for run_mode
+        in ['live', 'skip']
+    }
+
+    skip_mode_validate(taskdefs)
+
+    message = caplog.messages[0]
+
+    assert 'skip mode:\n    * skip_task' in message
+    assert ' live mode' not in message   # Avoid matching "non-live mode"
+    assert 'workflow mode' not in message
```

Comment:
I would suggest testing for exactly what is expected, and avoid testing for the absence of strings if possible as that's fragile.
```suggestion
    assert len(caplog.records) == 1
    assert log_filter(
        level=logging.WARNING,
        exact_match=(
            "The following tasks are set to run in skip mode:\n"
            "    * skip_task"
        ),
        log=caplog
    )
```